cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/adapter/AdapterService.java

```diff
@@ -0,0 +1,514 @@
+      }
+    } catch (DatasetManagementException e) {
+      LOG.error("Error while creating dataset {}", datasetName, e);
+      throw new RuntimeException(e);
+    } catch (IOException e) {
+      LOG.error("Error while creating dataset {}", datasetName, e);
+      throw new RuntimeException(e);
+    }
+  }
+
+  // Reads all the jars from the adapter directory and sets up required internal structures.
+  @VisibleForTesting
+  void registerAdapters() {
+    try {
+      File baseDir = new File(configuration.get(Constants.AppFabric.ADAPTER_DIR));
+      Collection<File> files = FileUtils.listFiles(baseDir, new String[]{"jar"}, true);
+      for (File file : files) {
+        try {
+          Manifest manifest = new JarFile(file.getAbsolutePath()).getManifest();
+          AdapterTypeInfo adapterTypeInfo = createAdapterTypeInfo(file, manifest);
+          if (adapterTypeInfo != null) {
+            adapterTypeInfos.put(adapterTypeInfo.getType(), adapterTypeInfo);
+          } else {
+            LOG.error("Missing required information to create adapter {}", file.getAbsolutePath());
+          }
+        } catch (IOException e) {
+          LOG.warn(String.format("Unable to read adapter jar %s", file.getAbsolutePath()));
+        }
+      }
+    } catch (Exception e) {
+      LOG.warn("Unable to read the plugins directory");
+    }
+  }
+
+
+  private AdapterTypeInfo createAdapterTypeInfo(File file, Manifest manifest) {
+    if (manifest != null) {
+      Attributes mainAttributes = manifest.getMainAttributes();
+
+      String adapterType = mainAttributes.getValue(AdapterManifestAttributes.ADAPTER_TYPE);
+      String adapterProgramType = mainAttributes.getValue(AdapterManifestAttributes.ADAPTER_PROGRAM_TYPE);
+      String defaultAdapterProperties = mainAttributes.getValue(AdapterManifestAttributes.ADAPTER_PROPERTIES);
+
+      String sourceType = mainAttributes.getValue(AdapterManifestAttributes.SOURCE_TYPE);
+      String sinkType = mainAttributes.getValue(AdapterManifestAttributes.SINK_TYPE);
+      String defaultSourceProperties = mainAttributes.getValue(AdapterManifestAttributes.SOURCE_PROPERTIES);
+      String defaultSinkProperties = mainAttributes.getValue(AdapterManifestAttributes.SINK_PROPERTIES);
+
+      if (adapterType != null && sourceType != null && sinkType != null && adapterProgramType != null) {
+        return new AdapterTypeInfo(file, adapterType, Source.Type.valueOf(sourceType.toUpperCase()),
+                                   Sink.Type.valueOf(sinkType.toUpperCase()),
+                                   propertiesFromString(defaultSourceProperties),
+                                   propertiesFromString(defaultSinkProperties),
+                                   propertiesFromString(defaultAdapterProperties),
+                                   ProgramType.valueOf(adapterProgramType.toUpperCase()));
+      }
+    }
+    return null;
+  }
+
+  protected Map<String, String> propertiesFromString(String gsonEncodedMap) {
```

Comment:
I guess this works just to get things working for now, but it seems really messy to have values that are constants in the code as part of some json in the manifest file. Not sure what we'd want to do, but seems like we'd want to have some hooks for the adapters to supply default properties for sources and sinks that get created. Can we create a jira and reference it here